src/vmm/src/persist.rs

```diff
@@ -548,22 +583,48 @@ fn guest_memory_from_uffd(
         .create()
         .map_err(GuestMemoryFromUffdError::Create)?;
 
-    for mem_region in guest_memory.iter() {
-        uffd.register(mem_region.as_ptr().cast(), mem_region.size() as _)
-            .map_err(GuestMemoryFromUffdError::Register)?;
-    }
+    let fds = match guest_memfd_copy {
+        Some(ref guest_memfd) => {
+            guest_memory_uffd_register_minor(&guest_memory, &uffd);
+            vec![
+                uffd.as_raw_fd(),
+                guest_memfd.as_raw_fd(),
+                userfault_bitmap_memfd
+                    .expect("memfd is not present")
+                    .as_raw_fd(),
+            ]
+        }
+        None => {
+            for mem_region in guest_memory.iter() {
+                uffd.register(mem_region.as_ptr().cast(), mem_region.size() as _)
+                    .map_err(GuestMemoryFromUffdError::Register)?;
+            }
+            vec![uffd.as_raw_fd()]
+        }
+    };
 
-    send_uffd_handshake(mem_uds_path, &backend_mappings, &uffd)?;
+    // We prevent Rust from closing the guest_memfd file descriptor
+    // so the UFFD handler can used it to populate guest memory.
+    forget(guest_memfd_copy);
```

Comment:
since we only send over the raw FD number, we can just hoist `guest_memfd.as_raw_fd()` to the beginning of the function to store the fd number itself for sending across, before we pass ownership  of the fd to `create_memory`. Then there's no need for clone()/forget()